core/resource/handler/v1beta1/resource.go

```diff
@@ -0,0 +1,341 @@
+}
+
+func (rh ResourceHandler) CreateResource(ctx context.Context, req *pb.CreateResourceRequest) (*pb.CreateResourceResponse, error) {
+	tnnt, err := tenant.NewTenant(req.GetProjectName(), req.GetNamespaceName())
+	if err != nil {
+		return nil, errors.GRPCErr(err, "failed to create resource")
+	}
+
+	store, err := resource.FromStringToStore(req.GetDatastoreName())
+	if err != nil {
+		return nil, errors.GRPCErr(err, "invalid create resource request")
+	}
+
+	res, err := fromResourceProto(req.Resource, tnnt, store)
+	if err != nil {
+		return nil, errors.GRPCErr(err, "failed to create resource")
+	}
+
+	err = rh.service.Create(ctx, res)
+	if err != nil {
+		return nil, errors.GRPCErr(err, "failed to create resource "+res.FullName())
+	}
+	return &pb.CreateResourceResponse{}, nil
+}
+
+func (rh ResourceHandler) ReadResource(ctx context.Context, req *pb.ReadResourceRequest) (*pb.ReadResourceResponse, error) {
+	if req.GetResourceName() == "" {
+		return nil, errors.GRPCErr(errors.InvalidArgument(resource.EntityResource, "empty resource name"), "invalid read resource request")
+	}
+
+	store, err := resource.FromStringToStore(req.GetDatastoreName())
+	if err != nil {
+		return nil, errors.GRPCErr(err, "invalid read resource request")
+	}
+
+	tnnt, err := tenant.NewTenant(req.GetProjectName(), req.GetNamespaceName())
+	if err != nil {
+		return nil, errors.GRPCErr(err, "failed to read resource "+req.GetResourceName())
+	}
+
+	response, err := rh.service.Get(ctx, tnnt, store, req.GetResourceName())
+	if err != nil {
+		return nil, errors.GRPCErr(err, "failed to read resource "+req.GetResourceName())
+	}
+
+	protoResource, err := toResourceProto(response)
+	if err != nil {
+		return nil, errors.GRPCErr(err, "failed to read resource "+req.GetResourceName())
+	}
+
+	return &pb.ReadResourceResponse{
+		Resource: protoResource,
+	}, nil
+}
+
+func (rh ResourceHandler) UpdateResource(ctx context.Context, req *pb.UpdateResourceRequest) (*pb.UpdateResourceResponse, error) {
+	tnnt, err := tenant.NewTenant(req.GetProjectName(), req.GetNamespaceName())
+	if err != nil {
+		return nil, errors.GRPCErr(err, "failed to update resource")
+	}
+
+	store, err := resource.FromStringToStore(req.GetDatastoreName())
+	if err != nil {
+		return nil, errors.GRPCErr(err, "invalid update resource request")
+	}
+
+	res, err := fromResourceProto(req.Resource, tnnt, store)
+	if err != nil {
+		return nil, errors.GRPCErr(err, "failed to update resource")
+	}
+
+	err = rh.service.Update(ctx, res)
+	if err != nil {
+		return nil, errors.GRPCErr(err, "failed to update resource "+res.FullName())
+	}
+	return &pb.UpdateResourceResponse{}, nil
+}
+
+func writeError(logWriter writer.LogWriter, err error) {
+	if err == nil {
+		return
+	}
+	var me *errors.MultiError
+	if errors.As(err, &me) {
+		for _, e := range me.Errors {
+			writeError(logWriter, e)
+		}
+	} else {
+		var de *errors.DomainError
+		if errors.As(err, &de) {
+			logWriter.Write(writer.LogLevelError, de.DebugString())
```

Comment:
Debug string should be printed only on the server side. we should not send this to the client side. It will contain errors from external system.

Reply:
I agree that external error messages will be included here. However, if we don't send those messages, the user will get so minimal error message which I'm sure they will get confused on what is wrong with their request. Especially when they do deployment, if there are some migrations that involve many resources to be created or updated, it will be tricky for them to understand the encountered errors.